sdk/src/types/block/slot/commitment_id.rs

```diff
@@ -1,17 +1,116 @@
 // Copyright 2023 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-use super::SlotIndex;
+use crate::types::block::{slot::SlotIndex, ConvertTo, Error};
 
-impl_id!(pub SlotCommitmentId, 36, "Identifier of a slot commitment.");
+impl_id!(pub SlotCommitmentHash, 32, "Identifier of a slot commitment.");
 
-#[cfg(feature = "serde")]
-string_serde_impl!(SlotCommitmentId);
+impl SlotCommitmentHash {
+    #[cfg(target_endian = "little")]
+    pub fn with_slot_index(self, slot_index: SlotIndex) -> SlotCommitmentId {
+        SlotCommitmentId { hash: self, slot_index }
+    }
+
+    #[cfg(target_endian = "big")]
+    pub fn with_slot_index(self, slot_index: SlotIndex) -> SlotCommitmentId {
+        SlotCommitmentId {
+            hash: self,
+            slot_index: slot_index.to_le().into(),
+        }
+    }
+}
+
+/// A block identifier.
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Ord, PartialOrd, Debug, packable::Packable)]
+#[packable(unpack_error = Error)]
+#[repr(C)]
+pub struct SlotCommitmentId {
+    pub(crate) hash: SlotCommitmentHash,
+    // IMPORTANT: On big-endian systems this value is misrepresented because it is transmuted directly
+    // from bytes, so the getter below handles that conversion. Do not access it directly.
+    slot_index: SlotIndex,
+}
 
 impl SlotCommitmentId {
+    /// The length of a [`SlotCommitmentId`]
+    pub const LENGTH: usize = 36;
+
+    pub fn new(bytes: [u8; Self::LENGTH]) -> Self {
+        unsafe { core::mem::transmute(bytes) }
+    }
+
+    /// Returns the [`SlotCommitmentId`]'s hash part.
+    pub fn hash(&self) -> &SlotCommitmentHash {
+        &self.hash
+    }
+
+    /// Returns the [`SlotCommitmentId`]'s slot index part.
+    #[cfg(target_endian = "little")]
+    pub fn slot_index(&self) -> SlotIndex {
+        self.slot_index
+    }
+
     /// Returns the [`SlotCommitmentId`]'s slot index part.
+    #[cfg(target_endian = "big")]
     pub fn slot_index(&self) -> SlotIndex {
-        // Unwrap is fine because size is already known and valid.
-        SlotIndex::new(u32::from_le_bytes(self.0[32..36].try_into().unwrap()))
+        self.slot_index.to_le().into()
+    }
+}
+
+impl AsRef<[u8]> for SlotCommitmentId {
+    fn as_ref(&self) -> &[u8] {
+        unsafe { core::mem::transmute::<_, &[u8; Self::LENGTH]>(self) }
+    }
+}
```

Comment:
A comment wit the `bech32` crate devs made me realize that we probably the same situation here why they would not impl `AsRef<[u8]>` on certain types that both have a valid byte and a valid str representation. For IDs it's basically the same and it might be better to have dedicated methods `as_bytes` and `as_str` instead of `as_ref` which might be confusing. Probably worth an issue?

Reply:
The difference is that HRP is a text-first type so you have UTF8/ASCII issues. IDs are byte-first types so we don't. I think `AsRef<[u8]>` is fine 🤷🏻‍♂️

Reply:
Also tbh I kinda disagree with the bech32 crate people, I would have still added `AsRef<[u8]>`, it's super common.